// playwright/lib/server/firefox.js
"use strict";
/**
 * Copyright 2017 Google Inc. All rights reserved.
 * Modifications copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const os = require("os");
const fs = require("fs");
const path = require("path");
const ws = require("ws");
const ffBrowser_1 = require("../firefox/ffBrowser");
const ffConnection_1 = require("../firefox/ffConnection");
const helper_1 = require("../helper");
const browserServer_1 = require("./browserServer");
const browserType_1 = require("./browserType");
const transport_1 = require("../transport");
const logger_1 = require("../logger");
class Firefox extends browserType_1.BrowserTypeBase {
    constructor(packagePath, browser) {
        const webSocketRegex = /^Juggler listening on (ws:\/\/.*)$/;
        super(packagePath, browser, { webSocketRegex, stream: 'stdout' });
    }
    _connectToTransport(transport, options) {
        return ffBrowser_1.FFBrowser.connect(transport, options);
    }
    _amendEnvironment(env, userDataDir, executable, browserArguments) {
        return os.platform() === 'linux' ? {
            ...env,
            // On linux Juggler ships the libstdc++ it was linked against.
            LD_LIBRARY_PATH: `${path.dirname(executable)}:${process.env.LD_LIBRARY_PATH}`,
        } : env;
    }
    _attemptToGracefullyCloseBrowser(transport) {
        const message = { method: 'Browser.close', params: {}, id: ffConnection_1.kBrowserCloseMessageId };
        transport.send(message);
    }
    _wrapTransportWithWebSocket(transport, logger, port) {
        return wrapTransportWithWebSocket(transport, logger, port);
    }
    _defaultArgs(options, isPersistent, userDataDir) {
        const { devtools, headless } = browserType_1.processBrowserArgOptions(options);
        const { args = [], proxy } = options;
        if (devtools)
            console.warn('devtools parameter is not supported as a launch argument in Firefox. You can launch the devtools window manually.');
        const userDataDirArg = args.find(arg => arg.startsWith('-profile') || arg.startsWith('--profile'));
        if (userDataDirArg)
            throw new Error('Pass userDataDir parameter instead of specifying -profile argument');
        if (args.find(arg => arg.startsWith('-juggler')))
            throw new Error('Use the port parameter instead of -juggler argument');
        if (proxy) {
            options.firefoxUserPrefs = options.firefoxUserPrefs || {};
            options.firefoxUserPrefs['network.proxy.type'] = 1;
            const proxyServer = new URL(proxy.server);
            const isSocks = proxyServer.protocol === 'socks5:';
            if (isSocks) {
                options.firefoxUserPrefs['network.proxy.socks'] = proxyServer.hostname;
                options.firefoxUserPrefs['network.proxy.socks_port'] = parseInt(proxyServer.port, 10);
            }
            else {
                options.firefoxUserPrefs['network.proxy.http'] = proxyServer.hostname;
                options.firefoxUserPrefs['network.proxy.http_port'] = parseInt(proxyServer.port, 10);
                options.firefoxUserPrefs['network.proxy.ssl'] = proxyServer.hostname;
                options.firefoxUserPrefs['network.proxy.ssl_port'] = parseInt(proxyServer.port, 10);
            }
            if (proxy.bypass)
                options.firefoxUserPrefs['network.proxy.no_proxies_on'] = proxy.bypass;
        }
        if (options.firefoxUserPrefs) {
            const lines = [];
            for (const [name, value] of Object.entries(options.firefoxUserPrefs))
                lines.push(`user_pref(${JSON.stringify(name)}, ${JSON.stringify(value)});`);
            fs.writeFileSync(path.join(userDataDir, 'user.js'), lines.join('\n'));
        }
        const firefoxArguments = ['-no-remote'];
        if (headless) {
            firefoxArguments.push('-headless');
        }
        else {
            firefoxArguments.push('-wait-for-browser');
            firefoxArguments.push('-foreground');
        }
        firefoxArguments.push(`-profile`, userDataDir);
        firefoxArguments.push('-juggler', '0');
        firefoxArguments.push(...args);
        if (isPersistent)
            firefoxArguments.push('about:blank');
        else
            firefoxArguments.push('-silent');
        return firefoxArguments;
    }
}
exports.Firefox = Firefox;
function wrapTransportWithWebSocket(transport, logger, port) {
    const server = new ws.Server({ port });
    const guid = helper_1.helper.guid();
    const idMixer = new transport_1.SequenceNumberMixer();
    const pendingBrowserContextCreations = new Set();
    const pendingBrowserContextDeletions = new Map();
    const browserContextIds = new Map();
    const sessionToSocket = new Map();
    const sockets = new Set();
    transport.onmessage = message => {
        if (typeof message.id === 'number') {
            // Process command response.
            const seqNum = message.id;
            const value = idMixer.take(seqNum);
            if (!value)
                return;
            const { id, socket } = value;
            if (socket.readyState === ws.CLOSING) {
                if (pendingBrowserContextCreations.has(id)) {
                    transport.send({
                        id: ++transport_1.SequenceNumberMixer._lastSequenceNumber,
                        method: 'Browser.removeBrowserContext',
                        params: { browserContextId: message.result.browserContextId }
                    });
                }
                return;
            }
            if (pendingBrowserContextCreations.has(seqNum)) {
                // Browser.createBrowserContext response -> establish context attribution.
                browserContextIds.set(message.result.browserContextId, socket);
                pendingBrowserContextCreations.delete(seqNum);
            }
            const deletedContextId = pendingBrowserContextDeletions.get(seqNum);
            if (deletedContextId) {
                // Browser.removeBrowserContext response -> remove context attribution.
                browserContextIds.delete(deletedContextId);
                pendingBrowserContextDeletions.delete(seqNum);
            }
            message.id = id;
            socket.send(JSON.stringify(message));
            return;
        }
        // Process notification response.
        const { method, params, sessionId } = message;
        if (sessionId) {
            const socket = sessionToSocket.get(sessionId);
            if (!socket || socket.readyState === ws.CLOSING) {
                // Drop unattributed messages on the floor.
                return;
            }
            socket.send(JSON.stringify(message));
            return;
        }
        if (method === 'Browser.attachedToTarget') {
            const socket = browserContextIds.get(params.targetInfo.browserContextId);
            if (!socket || socket.readyState === ws.CLOSING) {
                // Drop unattributed messages on the floor.
                return;
            }
            sessionToSocket.set(params.sessionId, socket);
            socket.send(JSON.stringify(message));
            return;
        }
        if (method === 'Browser.detachedFromTarget') {
            const socket = sessionToSocket.get(params.sessionId);
            sessionToSocket.delete(params.sessionId);
            if (socket && socket.readyState !== ws.CLOSING)
                socket.send(JSON.stringify(message));
            return;
        }
    };
    transport.onclose = () => {
        for (const socket of sockets) {
            socket.removeListener('close', socket.__closeListener);
            socket.close(undefined, 'Browser disconnected');
        }
        server.close();
        transport.onmessage = undefined;
        transport.onclose = undefined;
    };
    server.on('connection', (socket, req) => {
        if (req.url !== '/' + guid) {
            socket.close();
            return;
        }
        sockets.add(socket);
        socket.on('message', (message) => {
            const parsedMessage = JSON.parse(Buffer.from(message).toString());
            const { id, method, params } = parsedMessage;
            const seqNum = idMixer.generate({ id, socket });
            transport.send({ ...parsedMessage, id: seqNum });
            if (method === 'Browser.createBrowserContext')
                pendingBrowserContextCreations.add(seqNum);
            if (method === 'Browser.removeBrowserContext')
                pendingBrowserContextDeletions.set(seqNum, params.browserContextId);
        });
        socket.on('error', logger_1.logError(logger));
        socket.on('close', socket.__closeListener = () => {
            for (const [browserContextId, s] of browserContextIds) {
                if (s === socket) {
                    transport.send({
                        id: ++transport_1.SequenceNumberMixer._lastSequenceNumber,
                        method: 'Browser.removeBrowserContext',
                        params: { browserContextId }
                    });
                    browserContextIds.delete(browserContextId);
                }
            }
            sockets.delete(socket);
        });
    });
    const address = server.address();
    const wsEndpoint = typeof address === 'string' ? `${address}/${guid}` : `ws://127.0.0.1:${address.port}/${guid}`;
    return new browserServer_1.WebSocketWrapper(wsEndpoint, [pendingBrowserContextCreations, pendingBrowserContextDeletions, browserContextIds, sessionToSocket, sockets]);
}
